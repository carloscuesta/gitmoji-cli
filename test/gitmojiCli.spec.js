/* global describe, it, expect */
const GitmojiCli = require('../src/gitmoji')
const config = require('../src/config')
const constants = require('../src/constants')
const guard = require('../src/guard')
const prompts = require('../src/prompts')
const utils = require('../src/utils')
const stubs = require('./stubs')

const gitmojiCli = new GitmojiCli(stubs.gitmojiApiClient)

describe('constants module', () => {
  it('should match for the exported constants', () => {
    expect(constants).toMatchSnapshot()
  })
})

describe('prompts module', () => {
  it('should match for the config prompts', () => {
    expect(prompts.config).toMatchSnapshot()
  })

  it('should match for the gitmoji prompts', () => {
    const gitmojiPrompt = prompts.gitmoji(stubs.gitmojis, 'code', 'github')
    expect(prompts.gitmoji(gitmojiPrompt)).toMatchSnapshot()
  })

  it('should commit using the text emoji', () => {
    config.setEmojiFormat('code')
    const question = prompts.gitmoji(stubs.gitmojis, 'code', 'github')[0]
    return question.source(null, 'zap').then((emojis) => {
      expect(emojis[0].value).toMatchSnapshot()
    })
  })

  it('should commit using the unicode emoji', () => {
    config.setEmojiFormat('emoji')
    const question = prompts.gitmoji(stubs.gitmojis, 'emoji', 'github')[0]
    return question.source(null, 'zap').then((emojis) => {
      expect(emojis[0].value).toMatchSnapshot()
    })
  })
})

describe('config module', () => {
  it('should match the exported methods', () => {
    expect(config).toMatchSnapshot()
  })

  it('should match for setAutoAdd and getAutoAdd', () => {
    config.setAutoAdd(false)
    expect(config.getAutoAdd()).toMatchSnapshot()
  })

  it('should match for setEmojiFormat and getEmojiFormat', () => {
    config.setEmojiFormat('code')
    expect(config.getEmojiFormat()).toMatchSnapshot()
  })

  it('should match for setIssueFormat and getIssueFormat', () => {
    config.setIssueFormat('github')
    expect(config.getIssueFormat()).toMatchSnapshot()
  })

  it('should match for setSignedCommit and getSignedCommit', () => {
    config.setSignedCommit(false)
    expect(config.getSignedCommit()).toMatchSnapshot()
  })
})

describe('gitmoji module', () => {
  it('should match for gitmoji class', () => {
    expect(gitmojiCli).toMatchSnapshot()
  })

  describe('commit', () => {
    it('should match for the commit snapshot with the given prompts', () => {
      config.setIssueFormat('github')
      config.setSignedCommit(true)
      expect(gitmojiCli._commit(stubs.prompts)).toMatchSnapshot()
    })

    it('should match for the commit snapshot with the given prompts', () => {
      config.setIssueFormat('jira')
      config.setSignedCommit(false)
      expect(gitmojiCli._commit(stubs.promptsJira)).toMatchSnapshot()
    })
  })

  describe('_isAGitRepo', () => {
    it('should return true if a git repo is found', () => {
      expect(gitmojiCli._isAGitRepo()).toBe(true)
    })
  })
})

describe('utils module', () => {
  describe('findGitmojiCommand', () => {
    stubs.commands.map((command) => {
      it(`it should match ${command} command`, () => {
        utils.findGitmojiCommand(
          stubs.cliMock({ [command]: true }),
          stubs.optionsMock
        )
        expect(stubs.optionsMock[command]).toHaveBeenCalled()
      })
    })
  })

  describe('inputCountTransformer', () => {
    it('should match for an input with characters count string', () => {
      expect(
        utils.inputCountTransformer(stubs.prompts.title, stubs.titleMaxLength)
      ).toMatchSnapshot()
    })
  })
})

describe('guard module', () => {
  describe('title', () => {
    it('should return true when is valid', () => {
      expect(guard.title(stubs.prompts.title)).toBe(true)
    })

    it('should return error message when contains invalid characters', () => {
      expect(guard.title(stubs.invalidTitleMessageChar)).toMatchSnapshot()
    })

    it('should return an error message when its empty', () => {
      expect(guard.title(undefined)).toMatchSnapshot()
    })
  })

  describe('message', () => {
    it('should return true when is valid', () => {
      expect(guard.message(stubs.prompts.message)).toBe(true)
    })

    it('should return error message when contains invalid characters', () => {
      expect(guard.message(stubs.invalidTitleMessageChar)).toMatchSnapshot()
    })
  })

  describe('reference', () => {
    describe('default', () => {
      it('should return true if no reference is provided', () => {
        expect(guard.reference(undefined)).toBe(true)
      })

      it('should match return true when is valid github reference', () => {
        const assert = guard.reference(stubs.prompts.reference)
        expect(assert).toBe(true)
      })

      it('should return error message when contains invalid characters', () => {
        const assert = guard.reference(`#${stubs.prompts.reference}`)
        expect(assert).toMatchSnapshot()
      })
    })

    describe('github', () => {
      it('should return true when is valid', () => {
        const assert = guard.reference(
          stubs.prompts.reference,
          constants.GITHUB
        )
        expect(assert).toBe(true)
      })

      it('should return error message when contains invalid characters', () => {
        const assert = guard.reference(
          `#${stubs.prompts.reference}`,
          constants.GITHUB
        )
        expect(assert).toMatchSnapshot()
      })
    })

    describe('jira', () => {
      it('should return true when is valid', () => {
        const assert = guard.reference(
          stubs.promptsJira.reference,
          constants.JIRA
        )
        expect(assert).toBe(true)
      })

      it('should return error message when contains invalid characters', () => {
        const assert = guard.reference(
          `#${stubs.promptsJira.reference}`,
          constants.JIRA
        )
        expect(assert).toMatchSnapshot()
      })
    })
  })
})
