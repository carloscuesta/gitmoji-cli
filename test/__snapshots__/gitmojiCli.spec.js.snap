// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config module should match for setAutoAdd and getAutoAdd 1`] = `false`;

exports[`config module should match for setEmojiFormat and getEmojiFormat 1`] = `"code"`;

exports[`config module should match for setIssueFormat and getIssueFormat 1`] = `"github"`;

exports[`config module should match for setSignedCommit and getSignedCommit 1`] = `false`;

exports[`config module should match the exported methods 1`] = `
Object {
  "getAutoAdd": [Function],
  "getEmojiFormat": [Function],
  "getIssueFormat": [Function],
  "getSignedCommit": [Function],
  "setAutoAdd": [Function],
  "setEmojiFormat": [Function],
  "setIssueFormat": [Function],
  "setSignedCommit": [Function],
}
`;

exports[`constants module should match for the exported constants 1`] = `
Object {
  "AUTO_ADD": "autoAdd",
  "CODE": "code",
  "EMOJI_FORMAT": "emojiFormat",
  "GITHUB": "github",
  "HOOK_FILE_CONTENTS": "#!/bin/sh
# gitmoji as a commit hook
exec < /dev/tty
gitmoji --hook $1
",
  "HOOK_MODE": "hook",
  "HOOK_PATH": "/hooks/prepare-commit-msg",
  "HOOK_PERMISSIONS": 509,
  "ISSUE_FORMAT": "issueFormat",
  "JIRA": "jira",
  "SIGNED_COMMIT": "signedCommit",
  "TITLE_MAX_LENGTH_COUNT": 48,
}
`;

exports[`gitmoji module commit should match for the commit snapshot with the given prompts 1`] = `"git commit -S -m \\":zap: Improving performance issues.\\" -m \\"Refactored code. #5\\""`;

exports[`gitmoji module commit should match for the commit snapshot with the given prompts 2`] = `"git commit  -m \\":zap: Improving performance issues.\\" -m \\"Refactored code. ABC-123\\""`;

exports[`gitmoji module should match for gitmoji class 1`] = `
GitmojiCli {
  "_gitmojiApiClient": [Function],
  "_gitmojis": undefined,
}
`;

exports[`guard module message should return error message when contains invalid characters 1`] = `"[31mEnter a valid commit message[39m"`;

exports[`guard module reference default should return error message when contains invalid characters 1`] = `"[31mEnter the number of the reference without the #. Eg: 12[39m"`;

exports[`guard module reference github should return error message when contains invalid characters 1`] = `"[31mEnter the number of the reference without the #. Eg: 12[39m"`;

exports[`guard module reference jira should return error message when contains invalid characters 1`] = `"[31mEnter the JIRA reference key, such as ABC-123[39m"`;

exports[`guard module title should return an error message when its empty 1`] = `"[31mEnter a valid commit title[39m"`;

exports[`guard module title should return error message when contains invalid characters 1`] = `"[31mEnter a valid commit title[39m"`;

exports[`prompts module should commit using the text emoji 1`] = `":zap:"`;

exports[`prompts module should commit using the unicode emoji 1`] = `"⚡️"`;

exports[`prompts module should match for the config prompts 1`] = `
Array [
  Object {
    "message": "Enable automatic \\"git add .\\"",
    "name": "autoAdd",
    "type": "confirm",
  },
  Object {
    "choices": Array [
      "github",
      "jira",
    ],
    "message": "Choose Issue Format",
    "name": "issueFormat",
    "type": "list",
  },
  Object {
    "choices": Array [
      Object {
        "name": ":smile:",
        "value": "code",
      },
      Object {
        "name": "😄",
        "value": "emoji",
      },
    ],
    "message": "Select how emojis should be used in commits",
    "name": "emojiFormat",
    "type": "list",
  },
  Object {
    "message": "Enable signed commits",
    "name": "signedCommit",
    "type": "confirm",
  },
]
`;

exports[`prompts module should match for the gitmoji prompts 1`] = `
Array [
  Object {
    "message": "Choose a gitmoji:",
    "name": "gitmoji",
    "source": [Function],
    "type": "autocomplete",
  },
  Object {
    "message": "Enter the commit title",
    "name": "title",
    "transformer": [Function],
    "validate": [Function],
  },
  Object {
    "message": "Enter the commit message:",
    "name": "message",
    "validate": [Function],
  },
  Object {
    "message": "Issue / PR reference:",
    "name": "reference",
    "validate": [Function],
  },
]
`;

exports[`utils module inputCountTransformer should match for an input with characters count string 1`] = `"[29/48]: Improving performance issues."`;
